packages/block-editor/src/components/provider/use-block-entity-sync.js

```diff
@@ -0,0 +1,163 @@
+
+			// Avoid the more expensive equality check if there are no controlled
+			// inner block areas.
+			const areBlocksDifferent = areAnyInnerBlocksControlled()
+				? ! isShallowEqual( newBlocks, blocks )
+				: newBlocks !== blocks;
+
+			if (
+				areBlocksDifferent &&
+				( pendingChanges.current.incoming ||
+					__unstableIsLastBlockChangeIgnored() )
+			) {
+				pendingChanges.current.incoming = null;
+				blocks = newBlocks;
+				isPersistent = newIsPersistent;
+				return;
+			}
+
+			// Since we often dispatch an action to mark the previous action as
+			// persistent, we need to make sure that the blocks changed on the
+			// previous action before committing the change.
+			const didPersistenceChange =
+				previousAreBlocksDifferent &&
+				! areBlocksDifferent &&
+				newIsPersistent &&
+				! isPersistent;
+
+			if ( areBlocksDifferent || didPersistenceChange ) {
+				// When knowing the blocks value is changing, assign instance
+				// value to skip reset in subsequent `componentDidUpdate`.
+				if ( areBlocksDifferent ) {
+					pendingChanges.current.outgoing.push( newBlocks );
+				}
+
+				blocks = newBlocks;
+				isPersistent = newIsPersistent;
+
+				const selectionStart = getSelectionStart();
+				const selectionEnd = getSelectionEnd();
+
+				if ( isPersistent ) {
+					onChange( blocks, { selectionStart, selectionEnd } );
+				} else {
+					onInput( blocks, { selectionStart, selectionEnd } );
+				}
+			}
+			previousAreBlocksDifferent = areBlocksDifferent;
+		} );
+		return () => unsubscribe();
+	}, [ registry, onChange, onInput, clientId ] );
+
+	// Determine if blocks need to be reset when they change.
+	useEffect( () => {
+		if ( pendingChanges.current.outgoing.includes( controlledBlocks ) ) {
+			// Skip block reset if the value matches expected outbound sync
+			// triggered by this component by a preceding change detection.
+			// Only skip if the value matches expectation, since a reset should
+			// still occur if the value is modified (not equal by reference),
+			// to allow that the consumer may apply modifications to reflect
+			// back on the editor.
+			if (
+				last( pendingChanges.current.outgoing ) === controlledBlocks
+			) {
+				pendingChanges.current.outgoing = [];
+			}
+		} else {
+			// Reset changing value in all other cases than the sync described
+			// above. Since this can be reached in an update following an out-
+			// bound sync, unset the outbound value to avoid considering it in
+			// subsequent renders.
+			pendingChanges.current.outgoing = [];
+			pendingChanges.current.incoming = controlledBlocks;
+			setControlledBlocks();
+
+			if ( incomingSelectionStart && incomingSelectionEnd ) {
+				resetSelection( incomingSelectionStart, incomingSelectionEnd );
+			}
+		}
+	}, [ controlledBlocks ] );
+}
+
+export const withBlockEntitySync = createHigherOrderComponent(
+	( WrappedComponent ) => ( props ) => {
+		if ( props.value ) {
+			useBlockEntitySync( props );
+		}
```

Comment:
This is breaking the React hooks rules, we should  always call the hook, instead a `isEnabled` config for the hook might be needed.

Reply:
A better approach is to render two different components.

```jsx
function ControlledInput( { value, onChange, ...rest } ) {
	return <input value={ value } onChange={ onChange } { ...rest } />;
}

function UncontrolledInput( props ) {
	const [ value, onChange ] = useState();
	return <input { ...props } value={ value } onChange={ onChange } />;
}

function Input( props ) {
	if ( props.value && props.onChange ) {
		return <ControlledInput { ...props } />;
	}
	return <UncontrolledInput { ...props } />;
}
```

Reply:
Yeah, I was thinking about that, but wasn't sure how to solve the problem :/ The reason it "works" is because we don't expect `props.value` to ever change from truthy to falsey over the course of the lifecycle. (and so, hooks are always called in the same order, which I think is what the hook rules are getting at.) But yeah. I'll add two different components for block editor provider and for inner blocks.

Reply:
Why would you use `withBlockEntitySync` without `props.value`. I don't think we need the check at all.

Reply:
Unfortunately, without the check, it crashes in some scenarios. 🙃 The HOC format is always wrapped around InnerBlocks, which may or may not be uncontrolled. (i.e. `value` can be undefined`).

So you definitely wouldn't use `withBlockEntitySync` without `props.value`. So this check avoids using the hook when we don't have that value 🤷

Reply:
for the top level provider, I did not notice any issues without the check. But I had to add it when I implemented it with innerblocks

Reply:
Right, this is only needed while `InnerBlocks` is still a class component. Makes sense.

Reply:
I added a `ControlledInnerBlocks` component which wraps the hook and the `UncontrolledInnerBlocks` component. Let me know what you think